File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        long long maxProd = *max_element(nums.begin(), nums.end());
	    long long currMax = 1;
	    long long currMin = 1;
	    for(auto a: nums){
	        if(a == 0){
	            currMax = 1;
    	        currMin = 1;
	        }
	        else{
	            long long currMaxTemp = currMax;
	            long long currMinTemp = currMin;
	            currMax = max({a*currMaxTemp, a*currMinTemp, (long long)a});
	            currMin = min({a*currMaxTemp, a*currMinTemp, (long long)a});
	            maxProd = max(maxProd, currMax);
	        }
	    }
	    return maxProd;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
# class Solution:
#     def maxProduct(self, nums: List[int]) -> int:
#         def scan(arr):
#             curr_prod = 1
#             max_prod = float("-inf")

#             for num in arr:
#                 curr_prod *= num
#                 max_prod = max(max_prod, curr_prod)

#                 if num == 0:
#                     curr_prod = 1

#             return max_prod

#         return max(scan(nums), scan(nums[::-1]))


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        curr_max = curr_min = max_prod = nums[0]

        for num in nums[1:]:
            if num < 0:
                curr_max, curr_min = curr_min, curr_max
            
            curr_max = max(num, curr_max * num)
            curr_min = min(num, curr_min * num)
            max_prod = max(max_prod, curr_max)

        return max_prod